### DISCUSSION TITLE
PageCache storing headers in lower format

### DISCUSSION THREAD

**Discussion by nadar** [yiisoft/yii2#19186] [Q&A] [answered]
We are using PageCache to cache a REST response, it seems that headers are always returned as lowercase chars afterwards, but the original input is case sensitive.

https://github.com/yiisoft/yii2/blob/master/framework/filters/PageCache.php#L281

Whats the reason for this behavior? The default active data provider headers are f.e `X-Pagination-Page-Count` but when cached it returns `x-pagination-total-count` afterwards.

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/commit/ac623ab8f6004f26d1668234a22d562ece1b3df2 can't remember exact reason for it.

**Comment by nadar** [comment, marked as answer]
We are describing the headers in an openapi file, so this leaves us in a bad situation, because the header names changes after caching :-)

**Reply by samdark** [reply]
That could be considered as a bug to be fixed so please create an issue (and a pull request if you have some time). One thing is that there should be a note in UPGRADE since that may break apps.

**Reply by samdark** [reply]
https://github.com/yiisoft/yii2/issues/19187

**Reply by samdark** [reply]
@bizley, @rob006 maybe you remember the reason?

**Reply by nadar** [reply]
I will do some more research today, and make some more tests. I also found this, https://stackoverflow.com/a/5259004/4611030 i was not aware of that, but i even if this should considered as best practice, then all headers should be lowercase whether they are cached or not. What also makes me curious is the fact that i am the first person complain about this after 7 years, sounds unlikely to me for such an obvious "issue".

**Reply by bizley** [reply]
Yes, according to spec header names should be case insensitive unless something changed that I'm not aware of.

**Reply by nadar** [reply]
If a debug the [restoreResponse() function here](https://github.com/yiisoft/yii2/blob/master/framework/filters/PageCache.php#L219), i can see the lower case headers:

![image](https://user-images.githubusercontent.com/3417221/150747571-7ed99493-f174-4e68-a81f-ac8e7e2954a6.png)

But CURL, chrome devtools and even Postman return the headers ALL Case Sensitive... wow this confuses me even more now.

![image](https://user-images.githubusercontent.com/3417221/150747991-ba54fe13-e46f-4797-9749-d8bd287cec8c.png)

Do those tools camel case the headers for better readability....?

**Reply by bizley** [reply]
I think it's oldest typical story with specs everywhere - we can have them but for some folks these are just "suggestions". I would do the Symfony approach and lowercase them internally for processing but leave them as-is for storing/displaying/sending/outputting etc..

**Comment by bizley** [comment]
In the mentioned commit we can see:

> HTTP header names (case-insensitive) indicating which HTTP headers can be cached.
     * Note if your HTTP headers contain sensitive information, you should white-list which headers can be cached.

**Comment by bizley** [comment]
I can see that Symfony is lowercasing the headers for processing purposes but keeping them as-is when resending.

**Comment by bizley** [comment]
I'm looking into this and looks like we could store original header name along the normalized ones but I've got one question - if you are storing the same (case-sensitive wise) header several times just under the different case-style which one should be stored as a not normalized name? I would say the last one.
